plugins/reposync.py

```diff
@@ -100,36 +140,44 @@ def delete_old_local_packages(self, packages_to_download):
                             except OSError:
                                 logger.error(_("failed to delete file %s"), path)
 
-    def getcomps(self):
-        for repo in self.base.repos.iter_enabled():
-            comps_fn = repo.metadata._comps_fn
-            if comps_fn:
-                if not os.path.exists(repo.pkgdir):
-                    try:
-                        os.makedirs(repo.pkgdir)
-                    except IOError:
-                        logger.error(_("Could not make repository directory: %s"), repo.pkgdir)
-                        sys.exit(1)
-                dest = os.path.join(self._repo_base_path[repo.id], 'comps.xml')
-                dnf.yum.misc.decompress(comps_fn, dest=dest)
-                logger.info(_("comps.xml for repository %s saved"), repo.id)
+    def getcomps(self, repo):
+        comps_fn = repo.metadata._comps_fn
+        if comps_fn:
+            dest = os.path.join(self.repo_target(repo), 'comps.xml')
+            dnf.yum.misc.decompress(comps_fn, dest=dest)
+            logger.info(_("comps.xml for repository %s saved"), repo.id)
 
-    def run(self):
-        base = self.base
-        base.conf.keepcache = True
+    def download_metadata(self, repo):
+        repo_target = self.repo_target(repo)
+        repo._repo.downloadMetadata(repo_target)
+        return True
 
-        query = base.sack.query().available()
+    def get_pkglist(self, repo):
+        query = self.base.sack.query().available().filterm(reponame=repo.id)
         if self.opts.newest_only:
             query = query.latest()
         if self.opts.source:
-            query = query.filter(arch='src')
+            query.filterm(arch='src')
         elif self.opts.arches:
-            query = query.filter(arch=self.opts.arches)
+            query.filterm(arch=self.opts.arches)
+        return query
 
+    def download_packages(self, repo):
+        base = self.base
+        pkglist = self.get_pkglist(repo)
+        progress = base.output.progress
+        remote_pkgs, local_repository_pkgs = base._select_remote_pkgs(pkglist)
+        if remote_pkgs:
+            drpm = dnf.drpm.DeltaInfo(base.sack.query().installed(), progress, 0)
+            payloads = [RPMPayloadLocation(pkg, progress, self.pkg_download_path(pkg))
+                        for pkg in remote_pkgs]
+            base._download_remote_payloads(payloads, drpm, progress, None)
+        if local_repository_pkgs:
+            for pkg in local_repository_pkgs:
+                pkg_path = os.path.join(pkg.repo.pkgdir, pkg.location.lstrip("/"))
+                target_dir = os.path.dirname(self.pkg_download_path(pkg))
+                dnf.util.ensure_dir(target_dir)
+                shutil.copy(pkg_path, target_dir)
```

Comment:
Please correct me, but I see an issue here. If I good understand rpms are downloaded in single location, then they are copied into final destination. In case that two rpms with same name but different content in repo (different directory), they are unified (only second downloaded file is redistributed).